libfreerdp/core/multitransport.c

```diff
@@ -124,47 +133,72 @@ BOOL multitransport_client_send_response(rdpMultitransport* multitransport, HRES
 		return FALSE;
 	}
 
-	Stream_Write_UINT32(s, multitransport->requestId);         /* requestId (4 bytes) */
-	Stream_Write_UINT16(s, multitransport->requestedProtocol); /* requestedProtocol (2 bytes) */
-	Stream_Write_UINT16(s, multitransport->reserved);          /* reserved (2 bytes) */
-	Stream_Write(s, multitransport->securityCookie,
-	             sizeof(multitransport->securityCookie)); /* securityCookie (16 bytes) */
-	Stream_Write_UINT32(s, hr);
-	return rdp_send_message_channel_pdu(multitransport->rdp, s, SEC_TRANSPORT_REQ);
+	Stream_Write_UINT32(s, reqId); /* requestId (4 bytes) */
+	Stream_Write_UINT32(s, hr);    /* HResult (4 bytes) */
+	return rdp_send_message_channel_pdu(multi->rdp, s, SEC_TRANSPORT_RSP);
 }
 
-BOOL multitransport_server_recv_response(rdpMultitransport* multitransport, wStream* s, HRESULT* hr)
+state_run_t multitransport_recv_response(rdpMultitransport* multi, wStream* s)
 {
-	rdpMultitransport multi = { 0 };
+	WINPR_ASSERT(multi && multi->rdp);
+	WINPR_ASSERT(s);
 
-	WINPR_ASSERT(multitransport);
-	WINPR_ASSERT(hr);
+	rdpSettings* settings = multi->rdp->settings;
+	if (!settings->ServerMode)
+	{
+		WLog_ERR(TAG, "client is not expecting a multi-transport resp packet");
+		return STATE_RUN_FAILED;
+	}
 
-	if (!Stream_CheckAndLogRequiredLength(TAG, s, 28))
-		return FALSE;
+	if (!Stream_CheckAndLogRequiredLength(TAG, s, 8))
+		return STATE_RUN_FAILED;
 
-	Stream_Read_UINT32(s, multi.requestId);         /* requestId (4 bytes) */
-	Stream_Read_UINT16(s, multi.requestedProtocol); /* requestedProtocol (2 bytes) */
-	Stream_Read_UINT16(s, multi.reserved);          /* reserved (2 bytes) */
-	Stream_Read(s, multi.securityCookie,
-	            sizeof(multi.securityCookie)); /* securityCookie (16 bytes) */
-	Stream_Read_UINT32(s, *hr);
-	if (!multitransport_compare(multitransport, &multi))
-		return FALSE;
-	return TRUE;
+	UINT32 requestId;
+	HRESULT hr;
+
+	Stream_Read_UINT32(s, requestId); /* requestId (4 bytes) */
+	Stream_Read_UINT32(s, hr);        /* hrResponse (4 bytes) */
+
+	return IFCALLRESULT(STATE_RUN_SUCCESS, multi->MtResponse, multi, requestId, hr);
+}
+
+static state_run_t multitransport_no_udp(rdpMultitransport* multi, UINT32 reqId, UINT16 reqProto,
+                                         const BYTE* cookie)
+{
+	return multitransport_client_send_response(multi, reqId, E_ABORT) ? STATE_RUN_SUCCESS
+	                                                                  : STATE_RUN_FAILED;
+}
+
+static state_run_t multitransport_server_handle_response(rdpMultitransport* multi, UINT32 reqId,
+                                                         UINT32 hrResponse)
+{
+	rdpRdp* rdp = multi->rdp;
```

Comment:
`WINPR_ASSERT` input variables (used as function pointer, so `static` does not apply)